x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheService.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.searchablesnapshots.cache;
+
+import org.elasticsearch.common.UUIDs;
+import org.elasticsearch.common.cache.Cache;
+import org.elasticsearch.common.cache.CacheBuilder;
+import org.elasticsearch.common.component.AbstractLifecycleComponent;
+import org.elasticsearch.common.component.Lifecycle;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.ByteSizeUnit;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.common.util.concurrent.ReleasableLock;
+import org.elasticsearch.core.internal.io.IOUtils;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Predicate;
+
+/**
+ * {@link CacheService} maintains a cache entry for all files read from cached searchable snapshot directories (see {@link CacheDirectory})
+ */
+public class CacheService extends AbstractLifecycleComponent {
+
+    public static final Setting<ByteSizeValue> SNAPSHOT_CACHE_SIZE_SETTING = Setting.byteSizeSetting("xpack.searchable.snapshot.cache.size",
+        new ByteSizeValue(1, ByteSizeUnit.GB),                  // TODO: size the default value according to disk space
+        new ByteSizeValue(0, ByteSizeUnit.BYTES),               // min // NORELEASE
+        new ByteSizeValue(Long.MAX_VALUE, ByteSizeUnit.BYTES),  // max
+        Setting.Property.NodeScope);
+
+    private final ReleasableLock cacheInvalidationLock;
+    private final ReleasableLock cacheAccessLock;
+    private final Cache<String, CacheFile> cache;
+
+    public CacheService(final Settings settings) {
+        this.cache = CacheBuilder.<String, CacheFile>builder()
+            .setMaximumWeight(SNAPSHOT_CACHE_SIZE_SETTING.get(settings).getBytes())
+            .weigher((key, entry) -> entry.getLength())
+            // NORELEASE This does not immediately free space on disk, as cache file are only deleted when all index inputs
+            // are done with reading/writing the cache file
+            .removalListener(notification -> IOUtils.closeWhileHandlingException(() -> notification.getValue().startEviction()))
+            .build();
+
+        // Prevent new CacheFile objects to be added to the cache while the cache is being fully or partially invalidated
+        // This can happen because CacheFile objects might execute listeners at eviction time, potentially forcing more
+        // objects (like CacheDirectory's index inputs) to get new CacheFile.
```

Comment:
I don't follow this. We invalidate the cache when stopping and expected that in that case any attempts to get things from the cache would result in exceptions. Can you clarify?

Reply:
Hum, I'm confused too :confused: I remember I added this because of some kind of deadlock in integration tests (not in manual tests, though) where the full invalidation ran concurrently with some file accesses in Lucene directory (executed by nodes list shard stores actions). The invalidation blocked the cache but the file accesses tried to get/compute new entries in cache, resulting in deadlocks. I've ran more tests without these locks and everything is fine so I now think that the deadlocks were just a manifestation of some previous buggy code.